File: segment_colors.py
import cv2
import numpy as np


def preview(image, desc="test"):
    while 1:
        cv2.imshow(desc, image)
        if cv2.waitKey(10) & 0xFF == 27:
            break

    cv2.destroyAllWindows()


def pick_colors(image, n, desc="Choose colors"):
    global counter
    while counter < n:
        cv2.imshow(desc, image)
        cv2.setMouseCallback(desc, mouse_points)
        if cv2.waitKey(10) & 0xFF == 27:
            break

    cv2.destroyAllWindows()


def mouse_points(event, x, y, flags, params):
    global counter
    global clicked_coords_mat
    global im
    global picked_rgb
    # Left button click
    if event == cv2.EVENT_LBUTTONDOWN:
        print(counter)
        print(y, x)
        print(np.flip(im[y, x]))  # reversed(BGR) = RGB
        clicked_coords_mat[counter] = y, x
        cur_cols = []
        for i in range(y-1, y+2):
            if i < 0 or i == im.shape[0]:
                continue
            for j in range(x-1, x+2):
                if j < 0 or j == im.shape[1]:
                    continue
                cur_cols.append(np.flip(im[i, j]))
        cur_cols = np.vstack(cur_cols)
        picked_rgb[counter] = np.mean(cur_cols, axis=0)
        counter = counter + 1


def rgb_to_hsv(rgb_arr):
    r = rgb_arr[0]
    g = rgb_arr[1]
    b = rgb_arr[2]
    r, g, b = r / 255.0, g / 255.0, b / 255.0
    mx = max(r, g, b)
    mn = min(r, g, b)
    df = mx - mn
    if mx == mn:
        h = 0
    elif mx == r:
        h = (60 * ((g - b) / df) + 360) % 360
    elif mx == g:
        h = (60 * ((b - r) / df) + 120) % 360
    elif mx == b:
        h = (60 * ((r - g) / df) + 240) % 360
    if mx == 0:
        s = 0
    else:
        s = df / mx
    v = mx
    return h/2, s*255, v


def increase_brightness(img, value=30):
    hsv = cv2.cvtColor(img, cv2.COLOR_BGR2HSV)
    h, s, v = cv2.split(hsv)
    lim = 255 - value
    v[v > lim] = 255
    v[v <= lim] += value
    final_hsv = cv2.merge((h, s, v))
    img = cv2.cvtColor(final_hsv, cv2.COLOR_HSV2BGR)
    return img


def decrease_saturation(img, value=30):
    hsv = cv2.cvtColor(img, cv2.COLOR_BGR2HSV)
    h, s, v = cv2.split(hsv)
    lim = value
    s[s < lim] = 0
    s[s >= lim] -= value
    final_hsv = cv2.merge((h, s, v))
    img = cv2.cvtColor(final_hsv, cv2.COLOR_HSV2BGR)
    return img


# def decrease_saturation(img, value=)


if __name__ == "__main__":
    im = cv2.imread('images/test.jpg')
    # preview(im, desc="Original")
    # blur = cv2.blur(im, (7,7))

    n = 7
    clicked_coords_mat = np.zeros((n, 2), int)
    picked_rgb = np.zeros((n, 3), int)
    counter = 0

    # Manually sample colors (pick colors from pixels around it)
    pick_colors(im, n)
    # print(picked_rgb)
    picked_hsv = [rgb_to_hsv(x) for x in picked_rgb]
    # print(picked_hsv)
    hs = [x[0] for x in picked_hsv]
    low_h = min(hs) - 1 if min(hs) >= 1 else min(hs) + 178
    high_h = max(hs) + 1 if max(hs) < 179 else max(hs) - 179
    ss = [x[1] for x in picked_hsv]
    low_s = min(ss) - 5 if min(ss) >= 5 else 0
    high_s = max(ss) + 5 if max(ss) <= 250 else 255
    print(low_h, high_h)
    print(low_s, high_s)

    # convert the input image to hsv
    hsv_im = cv2.cvtColor(im, cv2.COLOR_BGR2HSV)
    # preview(hsv_im, desc="HSV")

    # hsv hue, saturation and brightness
    # hue int [0,179]; color spectrum think angle
    # saturation int [0,255]; basically color 0=gray
    # brightness int [0, 255]; 0=dark
    # define range of blue color in HSV
    # hvals = np.linspace(0,179,20)
    # for i in range(len(hvals)-1):
    mask = cv2.inRange(hsv_im, np.array([low_h, low_s, 20]), np.array([high_h, high_s, 255]))
    preview(mask, desc="mask init")
    kernel1 = cv2.getStructuringElement(cv2.MORPH_ELLIPSE, (2, 2))  # open
    kernel2 = cv2.getStructuringElement(cv2.MORPH_ELLIPSE, (15, 15))  # close
    kernel3 = cv2.getStructuringElement(cv2.MORPH_ELLIPSE, (15, 15))  # dilate
    # Open: erosion then dilation - to remove tiny spots else where
    mask = cv2.morphologyEx(mask, cv2.MORPH_OPEN, kernel1)
    preview(mask, desc="mask open")
    # Close: dilation then erosion - to remove tiny spots inside the selected region
    mask = cv2.morphologyEx(mask, cv2.MORPH_CLOSE, kernel2)
    preview(mask, desc="mask close")
    mask = cv2.dilate(mask, kernel3, iterations=1)
    preview(mask, desc="mask final")
    # create an inverse of the mask
    mask_inv = cv2.bitwise_not(mask)
    # Filter only the selected color from the original image
    # res = increase_saturation(im, value=50)
    res = cv2.bitwise_and(im, im, mask=mask)
    # preview(res)
    background = increase_brightness(im, value=75)
    background = decrease_saturation(background, value=50)
    gray_bg = cv2.cvtColor(background, cv2.COLOR_BGR2GRAY)
    background = cv2.bitwise_and(gray_bg, gray_bg, mask=mask_inv)
    # preview(background)
    # convert the one channelled grayscale background to a three channelled image
    background = np.stack((background,)*3, axis=-1)
    # super impose white by addWeighted()
    # add the foreground and the background
    out = cv2.add(res, background)
    preview(out)

File: segment_colors.cpp
#include <opencv2/opencv.hpp>
#include <iostream>

using namespace cv;
using namespace std;

void preview(Mat image, string desc = "test") {
    while (1) {
        imshow(desc, image);
        if (waitKey(10) & 0xFF == 27) {
            break;
        }
    }
    destroyAllWindows();
}

int counter;
Mat clicked_coords_mat;
Mat picked_rgb;

void pick_colors(Mat image, int n, string desc = "Choose colors") {
    counter = 0;
    clicked_coords_mat = Mat(n, 2, CV_32SC1, Scalar(0));
    picked_rgb = Mat(n, 3, CV_32SC1, Scalar(0));
    while (counter < n) {
        imshow(desc, image);
        setMouseCallback(desc, [](int event, int x, int y, int flags, void* userdata) {
            if (event == EVENT_LBUTTONDOWN) {
                int i = counter;
                int* clicked_coords_data = (int*)clicked_coords_mat.data;
                int* picked_rgb_data = (int*)picked_rgb.data;
                cout << i << endl;
                cout << y << " " << x << endl;
                Vec3b bgr = image.at<Vec3f>(y, x);
                cout << (int)bgr[2] << " " << (int)bgr[1] << " " << (int)bgr[0] << endl;
                clicked_coords_data[i * 2] = y;
                clicked_coords_data[i * 2 + 1] = x;
                vector<Vec3b> cur_cols;
                for (int j = y - 1; j <= y + 1; j++) {
                    if (j < 0 || j == image.rows) {
                        continue;
                    }
                    for (int k = x - 1; k <= x + 1; k++) {
                        if (k < 0 || k == image.cols) {
                            continue;
                        }
                        cur_cols.push_back(image.at<Vec3b>(j, k));
                    }
                }
                Mat(cur_cols).reshape(1, cur_cols.size()).convertTo(cur_cols, CV_32SC1);
                Mat mean_rgb = mean(cur_cols);
                picked_rgb_data[i * 3] = mean_rgb.at<int>(0);
                picked_rgb_data[i * 3 + 1] = mean_rgb.at<int>(1);
                picked_rgb_data[i * 3 + 2] = mean_rgb.at<int>(2);
                counter++;
            }
        }, NULL);
        if (waitKey(10) & 0xFF == 27) {
            break;
        }
    }
    destroyAllWindows();
}

vector<int> rgb_to_hsv(Vec3i rgb_arr) {
    float r = rgb_arr[0];
    float g = rgb_arr[1];
    float b = rgb_arr[2];
    r = r / 255.0;
    g = g / 255.0;
    b = b / 255.0;
    float mx = max(r, max(g, b));
    float mn = min(r, min(g, b));
    float df = mx - mn;
    float h, s, v;
    if (mx == mn) {
        h = 0;
    } else if (mx == r) {
        h = fmod(60 * ((g - b) / df) + 360, 360);
    } else if (mx == g) {
        h = fmod(60 * ((b - r) / df) + 120, 360);
    } else if (mx == b) {
        h = fmod(60 * ((r - g) / df) + 240, 360);
    }
    if (mx == 0) {
        s = 0;
    } else {
        s = df / mx;
    }
    v = mx;
    return h / 2, s * 255, v * 255;
}


cv::Mat increase_brightness(cv::Mat img, int value = 30) {
    cv::Mat hsv;
    cvtColor(img, hsv, cv::COLOR_BGR2HSV);
    std::vector<cv::Mat> hsv_split;
    cv::split(hsv, hsv_split);
    cv::Mat v_channel = hsv_split[2];

    int lim = 255 - value;
    cv::threshold(v_channel, v_channel, lim, 255, cv::THRESH_TOZERO_INV);
    v_channel = v_channel + value;

    cv::merge(hsv_split, hsv);
    cv::cvtColor(hsv, img, cv::COLOR_HSV2BGR);

    return img;
}


cv::Mat decrease_saturation(cv::Mat img, int value = 30) {
    cv::Mat hsv;
    cvtColor(img, hsv, cv::COLOR_BGR2HSV);
    std::vector<cv::Mat> hsv_split;
    cv::split(hsv, hsv_split);
    cv::Mat s_channel = hsv_split[1];

    int lim = value;
    cv::threshold(s_channel, s_channel, lim, 255, cv::THRESH_TOZERO);
    s_channel = s_channel - value;

    cv::merge(hsv_split, hsv);
    cv::cvtColor(hsv, img, cv::COLOR_HSV2BGR);

    return img;
}


// expect inputs
// path to an image file (this should be converted to GUID like id to point to image)
// number of clicks/touches (integer) to sample colors
int main(int argc, char* argv[]) {

    if (argc != 3) {
        cerr << "Usage: " << argv[0] << " [integer] [filename]" << endl;
        return 1;
    }

    int n = atoi(argv[1]);
    string fName = argv[2];  //"images/test.jpg"

    cv::Mat im = cv::imread(fName);
    cv::Mat clicked_coords_mat = cv::Mat::zeros(n, 2, CV_32S);
    cv::Mat picked_rgb = cv::Mat::zeros(n, 3, CV_32S);
    int counter = 0;

    pick_colors(im, n);

    std::vector<cv::Vec3f> picked_hsv;
    for (int i = 0; i < n; i++) {
        cv::Vec3b rgb = picked_rgb.at<cv::Vec3b>(i);
        cv::Vec3f hsv = rgb_to_hsv(rgb);
        picked_hsv.push_back(hsv);
    }

    float low_h, high_h, low_s, high_s;
    std::vector<float> hs, ss;
    for (auto hsv : picked_hsv) {
        hs.push_back(hsv[0]);
        ss.push_back(hsv[1]);
    }
    low_h = *std::min_element(hs.begin(), hs.end()) - 1;
    high_h = *std::max_element(hs.begin(), hs.end()) + 1;
    low_s = *std::min_element(ss.begin(), ss.end()) - 10;
    high_s = *std::max_element(ss.begin(), ss.end()) + 10;

    std::cout << low_h << " " << high_h << std::endl;
    std::cout << low_s << " " << high_s << std::endl;

    cv::Mat hsv_im;
    cvtColor(im, hsv_im, cv::COLOR_BGR2HSV);

    cv::Mat mask;
    cv::inRange(hsv_im, cv::Scalar(low_h, 50, 20), cv::Scalar(high_h, high_s, 255), mask);
    cv::Mat res;
    cv::bitwise_and(im, im, res, mask);

    cv::imshow("Result", res);
    cv::waitKey(0);
    cv::destroyAllWindows();

    return 0;
}




